core/lib/zksync_core/src/sync_layer/gossip/storage.rs

```diff
@@ -0,0 +1,351 @@
+        // ^ The number can get stale, but it's OK for our purposes
+        Self::block(&mut storage, miniblock_number)
+            .await?
+            .with_context(|| format!("Miniblock #{miniblock_number} disappeared from Postgres"))
+    }
+
+    async fn storage(&self) -> anyhow::Result<StorageProcessor<'_>> {
+        self.pool
+            .access_storage_tagged("sync_layer")
+            .await
+            .context("Failed to connect to Postgres")
+    }
+
+    async fn block(
+        storage: &mut StorageProcessor<'_>,
+        number: MiniblockNumber,
+    ) -> anyhow::Result<Option<FinalBlock>> {
+        let Some(block) = storage
+            .sync_dal()
+            .sync_block(number, Address::default(), true)
+            .await
+            .with_context(|| format!("Failed getting miniblock #{number} from Postgres"))?
+        else {
+            return Ok(None);
+        };
+        let block = sync_block_to_consensus_block(block)?;
+        Ok(Some(block))
+    }
+
+    async fn first_block(&self) -> anyhow::Result<FinalBlock> {
+        let mut storage = self.storage().await?;
+        Self::block(&mut storage, MiniblockNumber(0))
+            .await?
+            .context("Genesis miniblock not present in Postgres")
+    }
+
+    async fn sealed_miniblock_number(&self) -> anyhow::Result<BlockNumber> {
+        let mut storage = self.storage().await?;
+        let number = storage
+            .blocks_dal()
+            .get_sealed_miniblock_number()
+            .await
+            .context("Failed getting sealed miniblock number")?;
+        Ok(BlockNumber(number.0.into()))
+    }
+}
+
+#[async_trait]
+impl BlockStore for PostgresBlockStorage {
+    async fn head_block(&self, ctx: &ctx::Ctx) -> StorageResult<FinalBlock> {
+        tokio::select! {
+            () = ctx.canceled() => Err(ctx::Canceled.into()),
+            result = self.head_block() => result.map_err(StorageError::Database),
+        }
+    }
+
+    async fn first_block(&self, ctx: &ctx::Ctx) -> StorageResult<FinalBlock> {
+        tokio::select! {
+            () = ctx.canceled() => Err(ctx::Canceled.into()),
+            result = self.first_block() => result.map_err(StorageError::Database),
+        }
+    }
+
+    async fn last_contiguous_block_number(&self, ctx: &ctx::Ctx) -> StorageResult<BlockNumber> {
+        tokio::select! {
+            () = ctx.canceled() => Err(ctx::Canceled.into()),
+            result = self.sealed_miniblock_number() => result.map_err(StorageError::Database),
+        }
+    }
+
+    async fn block(
+        &self,
+        ctx: &ctx::Ctx,
+        number: BlockNumber,
+    ) -> StorageResult<Option<FinalBlock>> {
+        let get_block = async {
+            let number = u32::try_from(number.0).context("block number is too large")?;
+            let mut storage = self.storage().await?;
+            Self::block(&mut storage, MiniblockNumber(number)).await
+        };
+        tokio::select! {
```

Comment:
I think that at this point we should make `Ctx::wait` public (https://github.com/matter-labs/era-consensus/blob/ccc3ee909d6da397f94fc43898ae0df20cd5f94f/node/libs/concurrency/src/ctx/mod.rs#L196C4-L196C29) with a disclaimer that it should only be used on cancel-safe context-unaware futures.